javaslang/src/main/java/javaslang/collection/AbstractMap.java

```diff
@@ -47,6 +47,31 @@ public M put(Tuple2<? extends K, ? extends V> entry) {
 
     @SuppressWarnings("unchecked")
     @Override
+    public M putWith(K key, V value, BiFunction<? super V, ? super V, ? extends V> merge) {
+        Objects.requireNonNull(merge, "the merge function is null");
+        Option<V> currentValue = get(key);
+        if (currentValue.isEmpty()) {
+            return (M) put(key, value);
+        } else {
+            return (M) put(key, merge.apply(currentValue.get(), value));
+        }
+    }
+
+    @Override
+    public M putWith(Tuple2<? extends K, ? extends V> entry,
+                     BiFunction<? super V, ? super V, ? extends V> merge) {
```

Comment:
I've compared the signature with `Map.merge`:

```java
<U extends V> Map<K, V> merge(Map<? extends K, U> that,                  BiFunction<? super V, ? super U, ? extends V> collisionResolution);
              M         putWith(Tuple2<? extends K, ? extends V> entry,  BiFunction<? super V, ? super V, ? extends V> merge);
```

They are very similar. We should adjust `putWith` like this:

```java
<U extends V> M         putWith(Tuple2<? extends K, ? extends U> entry,  BiFunction<? super V, ? super U, ? extends V> merge);
```

That opens the type space of the second covariant argument of the `merge` function.

/cc @ruslansennov right?